python/ray/experimental/sgd/pytorch/pytorch_trainer.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import sys
+import torch
+
+import ray
+
+from ray.experimental.sgd.pytorch.pytorch_runner import PyTorchRunner
+from ray.experimental.sgd.pytorch import utils
+
+
+class PyTorchTrainer(object):
+    """Train a PyTorch model using distributed PyTorch"""
+
+    def __init__(
+            self,
+            model_creator,
+            data_creator,
+            optimizer_creator=None,  # SGD as default TODO: handle this
+            config=None,
+            num_replicas=1,
+            resources_per_replica=None,
+            batch_size=16,
+            backend="gloo"):
+        """Sets up the PyTorch trainer.
+
+            Args:
+                model_creator (dict -> torch.nn.Module): creates the model
+                    using the config.
+                data_creator (dict -> Dataset, Dataset): creates the training
+                    and validation data sets using the config.
+                optimizer_creator (model, dict -> loss, optimizer): creates the
+                    loss and optimizer using the config.
+                config (dict): configuration passed to 'model_creator',
+                    'data_creator', and 'optimizer_creator'.
+                batch_size (int): batch size used for SGD.
+                backend (string): backend used for distributed SGD. "gloo" or
+                    "nccl".
+        """
+        # TODO: add support for mixed precision
+        # TODO: add support for callbacks
+        if sys.platform == "darwin":
+            raise Exception((
+                "Distributed PyTorch is not supported on macOS. For more "
+                "information, see "
+                "https://github.com/pytorch/examples/issues/467."))
+
+        self.model_creator = model_creator
+        self.config = {} if config is None else config
+        self.optimizer_timer = utils.TimerStat(window_size=1)
+
+        if resources_per_replica is None:
+            resources_per_replica = utils.Resources(
+                num_cpus=0, num_gpus=0, resources={})
+        # TODO: support multiple GPUs
+        if resources_per_replica.num_gpus > 1:
+            raise ValueError("multi-GPU models are not supported yet")
```

Comment:
I don't think we need to support multi-gpu per replica?

Reply:
Perhaps not for pytorch, but yes for TF. @richardliaw are you suggesting removing the `"yet"`?

Reply:
Yeah - I was referring specifically to this class.

Reply:
Resolving this since mutli-GPU works if we configure PyTorch via `CUDA_VISIBLE_DEVICES`.